libsolidity/analysis/TypeChecker.cpp

```diff
@@ -814,11 +821,26 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 			if (!varType->canLiveOutsideStorage())
 				m_errorReporter.typeError(_variable.location(), "Type " + varType->toString() + " is only valid in storage.");
 	}
-	else if (
-		_variable.visibility() >= VariableDeclaration::Visibility::Public &&
-		!FunctionType(_variable).interfaceFunctionType()
-	)
-		m_errorReporter.typeError(_variable.location(), "Internal or recursive type is not allowed for public state variables.");
+	else if (_variable.visibility() >= VariableDeclaration::Visibility::Public)
+	{
+		FunctionType getter(_variable);
+		if (!_variable.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::ABIEncoderV2))
+		{
+			vector<string> unsupportedTypes;
+			for (auto const& param: getter.parameterTypes() + getter.returnParameterTypes())
+				if (!typeSupportedByOldABIEncoder(*param))
+					unsupportedTypes.emplace_back(param->toString());
+			if (!unsupportedTypes.empty())
+				m_errorReporter.typeError(
+					_variable.location(),
+					"The following types are only supported for getters in the new experimental ABI encoder. "
+					"Either remove \"public\" or use \"pragma experimental ABIEncoderV2;\" to enable the feature: " +
+					joinHumanReadable(unsupportedTypes)
+				);
+		}
+		if (!getter.interfaceFunctionType())
+			m_errorReporter.typeError(_variable.location(), "Internal or recursive type is not allowed for public state variables.");
```

Comment:
I think this could be tested a bit more. As far as I can see, we only have tests for internal, but not for recursive types in ``070_struct_accessor_one_array_only.sol`` and ``function_type_internal_public_variable.sol`` (these are the only tests that trigger this error.)

Reply:
But this could also be done in a separate PR :)